text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,402 @@
+```
+No lints reported.
+
+A non-rust-analyzer user could run the following to look for feedback:
+```console
+$ CARGO_BUILD_NITS=nit cargo clippy
+   Compiling foo
+note: creating a let-binding and then immediately returning it like `let x = expr; x` at the end of a block
+...
+    Finished `foo` profile [optimized + debuginfo] target(s) in 15.85s
+```
+
+They then go and post a PR.
+On the PR, the code above gets the following issue opened:
+```
+X Check failure
+
+Code scanning clippy
+
+creating a let-binding and then immediately returning it like `let x = expr; x` at the end of a block (warning)
+
+Show more details
+```
+This gives the contributor an opportunity to know of this potential improvement
+if they missed it and for the code reviewer to double check if they agreed with
+the author's determination to ignore it.
+
+For future PRs, Github should not report this as it should recognize that the report is for existing code.
+
+## Integrating with CI
+
+The above potential experience is dependent on being able to integrate linting with CI to report only "new lints" or "lints for changed code".
+
+For Github users, you can use [`clippy-sarif`](https://crates.io/crates/clippy-sarif) to report lints as SARIF alerts.
+Github [tracks these alerts across runs](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#providing-data-to-track-code-scanning-alerts-across-runs).
+
+These can also be implemented manually with any setup.
+For example, in a large, corporate, legacy C++ code base that [epage](https://github.com/epage/)
+worked with, a linter used a similar database to track what lints were "new" and reported those back to
+[ReviewBoard](https://www.reviewboard.org/).
+
+## Choosing lint levels
+
+When creating a lint or overriding a default lint level,
+its important to consider the semantics, something like:
+
+`error` is for hard errors that even prevent reporting of errors further in the build, running of tests, etc.
+This should be reserved for when there is no point building or testing anything further because of how broken the code is.
+
+`warn` generally has the semantics of a soft-error.
+Its non-blocking locally but will be blocked when the code is merged upstream.
+This should be used when eventual correctness or consistency is desired.
+
+`nit` is for coding suggestions for users that may or may not improve the code or may not need to be done yet.
+
+`allow` is for when there is no right answer (e.g. mutually exclusive lints), there are false positives, the lint is overly noisy, or the lint is of limited value.
+
+*Note:* This is for communicating the intent of this RFC and to get people started with this new feature and is not meant to precisely redefine the
+[lint levels](https://rustc-dev-guide.rust-lang.org/diagnostics.html#diagnostic-levels).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
```

Comment:
We might want to also add a `## Clippy` section here to commit to changes that need to be done in Clippy. For example, adding a new lint group and moving lints there, probably from the `pedantic` (`allow`->`nit`) and the `style` group (`warn`->`nit`).

Also opening this thread for @rust-lang/clippy  to discuss the implications this will have on Clippy.

Reply:
I also think some `restriction` lints can be raised to `nit` too, I see quite a few as "you can do this differently if you want to" (e.g. perhaps `as_underscore` or `deref_by_slicing`). Quite a few aren't strict at all and could usually be standard practice. At least for us, this category of lints is the most commonly raised to `warn` before committing.

The thing is, I for the most part already see the `pedantic` category as already doing both what this new `nit` group and what the current `nursery` group does. Is it perhaps best `pedantic` is raised to `nit`, and ones with FPs/are too strict are moved elsewhere? Though then those lints won't be enabled alongside `#![warn(clippy::pedantic)]`.

Reply:
Yeah actually we don't need "less than warn", we've got pedantic.

we just need a way to run pedantic lints on new code only and we'd be set.

Reply:
I'm fine calling out that clippy would also need to do any relevant work for the new lint level.  However, I would like to treat what new groups are needed, their naming, and any other lint organization related stuff as out of scope like setting lint levels.  I think it would be much easier, more productive conversation for you to do that within your team and through your team's decision making process for that level of work rather than pulling in more major decision making into this multi-team RFC.

Reply:
> Yeah actually we don't need "less than warn", we've got pedantic.
>
> we just need a way to run pedantic lints on new code only and we'd be set.

The differences between this and pedantic are that the user can move things into this level *and* its a cross-linter convention.

Reply:
I added a discussion on pedantic at 14faa694

Reply:
I added a clippy section in 2410bb1e

Reply:
> The differences between this and pedantic are that the user can move things into this level and its a cross-linter convention.

I don't really see that as a difference; the thing with it being in a group is that you can set its level wholesale, but there's no much benefit to being able to "move things into pedantic/nit" that's not gotten by moving things into allow/warn as needed.

That said I think supporting multiple lint level profiles would be very cool.

Reply:
>  I don't really see that as a difference; the thing with it being in a group is that you can set its level wholesale, but there's no much benefit to being able to "move things into pedantic/nit" that's not gotten by moving things into allow/warn as needed.

The end-user workflow, at least for a Cargo user, isn't great for moving things in and out of allow/warn for non-blocking lints.
- Users should have a static way (e.g. `[lints]`) for distinguishing between project-wide soft errors and nits.
- We need to control when to report these.  If cargo is filtering them, it needs to know under what criteria to filter them by.  If instead we are mucking with RUSTFLAGS under the hood, then a rebuild is needed to switch between the two modes.

Reply:
> The end-user workflow, at least for a Cargo user, isn't great for moving things in and out of allow/warn for non-blocking lints.

I don't understand: you use the Cargo lints table for this, yes? That can move between allow and warn pretty easily. Are you looking for a way to also do it over CLI?

(As I mention in a different comment near the end I think user-defined lint groupings, or lint profiles, for that  would be _great_)


> Users should have a static way (e.g. [lints]) for distinguishing between project-wide soft errors and nits.

I've noted this in another comment but I think the unstated assumption muddling things here is `-Dwarnings`. It sounds as if what you're looking for is a way to make further distinctions that _persist under `-Dwarnings`_.

Reply:
> However, I would like to treat what new groups are needed, their naming, and any other lint organization related stuff as out of scope like setting lint levels

Fully agree. I opened this thread so that the Clippy team has an isolate space in this RFC to discuss things. Those shouldn't necessarily end up in this RFC though.

Thanks for adding a Clippy section. Could you add a note to that section, that Clippy would make further decisions how to use this new feature in a team-specific decision making process?

Reply:
> Thanks for adding a Clippy section. Could you add a note to that section, that Clippy would make further decisions how to use this new feature in a team-specific decision making process?

I have the general disclaimer in the Summary along with another in the Motivation to remind people not to read too much into my lint level descriptions.  What are you hoping to get out of having more, particularly having them per-team?

Reply:
> I don't understand: you use the Cargo lints table for this, yes? That can move between allow and warn pretty easily. Are you looking for a way to also do it over CLI?

I think this is bleeding over into our other discussion at https://github.com/rust-lang/rfcs/pull/3730#discussion_r1851068850 which is about workflows